### DISCUSSION TITLE
ESP8266 resistive dimmer with SSR zero cross

### DISCUSSION THREAD

**Discussion by mpetcuRO** [arendst/Tasmota#21664] [Q&A]
I'm trying to create an DIY pool controller to command the pump/heater and monitor/automate from Home Assistant.
Using an ESP8266 will have the following features:
- power dimmer (with SSR Fotek SSR-100DA and Zero Cross detection using octocopter) - for a 3kw electric heater that I want to modulate power based on Solar production
- power monitoring (using PZEM-004T module)
- temp sensing (using DS18x20)
- On/Off plugs (using relay board)
- Tasmota configurations in pictures

Settings:
- counter 4 is enable and reading 100Hz (rectified mains, checked with oscilloscope)
- Tasmota sensors build version 14
- no other commands issues in Tasmota

I have the following questions:

1. The dimmer dos not work, (tested with an 170W incandescent bulb)
 - if so99 is disable  then bulb is pulsating when the dimmer slider is anywhere from 30% to 99%, below 30% bulb is off, above 99% bulb is on
 - if I enable so99, then the temperature sensors are not shown in the interface, and the bulb is on at 100% brightness at max position of slider, bulb is off if you move the slider anywhere in the range.
 2. Is there an incompatibility for so99 command and ds18x20 ?
![tasmota 1](https://github.com/arendst/Tasmota/assets/73108927/619ab39a-2982-4d22-8318-1c018f715ddd)
![tasmota 2](https://github.com/arendst/Tasmota/assets/73108927/20d9a0ae-78a1-4113-9b37-c412d8b6c9b8)

**Comment by lalo-uy** [comment]
SSR-100DA do not exist in oficial fotek datasheet.
Tha bigest is the 70DA, but as all of them have allready the zero cross
thriger incorporated. No need to check ac line. Just output a slow ( 2s ,
0.5 hz ) pwm signal.

Not all firmwares have all the opions compiled in. You may need to bulid
one yourself. Gitpod is way easy.

**Reply by mpetcuRO** [reply]
Hello, pretty new to Tasmota and I have no idea how to output that signal, anyway for me does not make sense as the microprocessor needs to know the position of the syne wave to open or close the relay to dim the power. There's not much info on the web on this subject using Tasmota.

I have ordered the SSR-60DA but after the order the shop called me and offered the 100DA do to missing stock, I presume all have the ZC circuit. (for resistive load is an extreme overkill but i work with what i find :) )
![IMG_20240620_164903](https://github.com/arendst/Tasmota/assets/73108927/e8b28241-2ee5-4d20-b625-fa2d0099735c)

**Comment by stefanbode** [comment]
Can you do a pwm 1 instead of 4. the dimmer uses always counter4 for the zerocross as shown in documentation. You can test the dimmer working properly by connecting an OLD light. Please no led. I used this either for development. I am not aware of sideeffects between ds1820 and the dimmer , but also never tested.

**Comment by mpetcuRO** [comment]
I already tested all combinations that I can think of, the PMW 4 was just left from one of the tests, is now back to 1.
here is a video if the symptoms:
- if os99 is off then PMW is 955Hz default, and the light bulb is pulsating randomly, I expect this as the 50hz of the mains are not in sync.
- if so99 is on then the PMW signal is 100hz that is the rectified 50hz and the microprocessor behaves normally you can see that in the video on the oscilloscope, I have tested with 50Hz (not rectified mains) and is the same the bub is off all the way to 100% PMW.

[VIDEO](https://youtu.be/I5WBFkg-6r0)

is there a way I can delay the PMW signal sent but still keep the zero cross reference so I can match if with the zero cross of the SSR?

**Reply by barbudor** [reply]
In zéro cross mode, you are supposed to get a zéro cross détection signal in the ESP to synchronize the timings
You don't seem to have such signal, do you ?

The zero cross mode is meant to be used with a dimmer like the DFrobot module, not with a SSR.

**Reply by stefanbode** [reply]
Ok, thanks for the video and that you have the oscilloscope shown. I do see with so99 on the PWM modulation is exactly what it should be. You can see that the pulse length changed with the dimmer setting. This looks all like it should be. So it is a electrical problem of the switching device. You can check if there is a change when going from trailing edge to leading edge. In Germany at least the grid company do not allow to dim devices with loads of 3kw. Creates To much noise in the system. But this is just a legal one. Technically it works. I do not see an error on the esp software side.

**Comment by mpetcuRO** [comment]
I have the counter 4, that is an octocopler with the rectified mains frequency.
Is there a way I can make a dimmer from an SSR that has zero cross circuit built in?
The issue is that I did not see anywhere an the internet a dimmer that can support the 3Kw load.

**Reply by barbudor** [reply]
Your bulb is 3kW ????????
Originally you were talking about controlling a heater
Controlling a heater and a buld has nothing to do

You don't control a heater with a high frequency PWM but with a low frequency with a period > 30 sec
You should use either PID or Thermostat mode, not PWM

**Comment by mpetcuRO** [comment]
Bulb is for testing.
The 3kw heater will be modulated based on solar production. 
An AC dimmer can do this to a bulb by dimming the bulb, the bulb will consume less wats in this case. I only want to consume the excess solar production to heat the pool water.

**Reply by barbudor** [reply]
Once again : you don't control a heater like a bulb
bulb => high frequency PWM
heater => low frequency
You are loosing your time trying to use zero-crossing and high frequency for a heater

**Reply by stefanbode** [reply]
@barbudor this time I do not agree. A heater is a normal linear resistor you can also trigger with a PWM signal. To be compliant with the grid company you're most time not allowed to do it. These devices have a specific dimmer on the count of half sine waves. E.g. doing every second to get 50%. But this we do not support so far.

**Reply by stefanbode** [reply]
You can also rectify to DC and do the ÜPWM on the DC side. Big capacitor are a must.

**Comment by joba-1** [comment]
I don’t think you are talking about the same thing.
Yes, technically you _can_ switch kW loads with PWM, but for good reasons you won’t.

Simply put: to make sure to avoid bad side effects like electromagnetic noise you need equipment that is much more expensive than the few cents you save by using the last few kWh for pool heating. 

Zero crossing is only near zero crossing and real world resistive load always has a capacitive or inductive component. These add up to non bearable levels in the kW range and to counter them you need measuring devices to fine tune more complicated circuits. Not practical for a self made excess energy pool heater.

**Comment by lalo-uy** [comment]
PWM is not restricted to 100 or more htz. I have a control of 16kw heater
with a 2 seconds pwm generated with a 555.  As the SSR is zero cross
triggered you get what is called whole cycles control.
It would be wonderful if Tasmota pwm could be set to very low freq.

**Reply by sfromis** [reply]
For a "PWM" frequency near 0.5 hz, you can config a relay, and use power blink to have it cycle. `BlinkTime` has a resolution of 100 ms.

**Comment by lalo-uy** [comment]
Since you have the opto hooked tho a counter, may be you could tourn the
power on every 1000 count and off at a desired proportional count.

**Reply by mpetcuRO** [reply]
Thanks for the ideas, I will test on my setup.
Sorry, but I do not see how a 2 sec signal will modulate the power of the heater, can you elaborate on your setup?

My idea was to modulate the power of the heater so in case I only have 1kw surplus of solar power to use only that.

**Reply by sfromis** [reply]
Such pulses can work fine if your purpose is to regulate how much heat is produced, but not to control how much power is drawn at any instant in time.

**Reply by lalo-uy** [reply]
The instant you switch on a 3kw R it will draw 3kw of power.
If yor solar system is providing only 1kw,  the output capacitor will be depleted .
Of its capacity will depend the pwm solution to work ok and its minimal freq.

**Comment by mpetcuRO** [comment]
ok, looks that this is way more complicated than I expected.
What will be the best technology to do this? TRIAC, SSR, Voltage regulator prebuilt that I can control with Tasmota.
I do not need for the adjustment to be linear,  if I have positions of 500W, 1000W, ... 3000W. I can still only consume the excess solar.
Can I just use something like this? [LINK](https://www.aliexpress.com/item/10000061653371.html)

**Comment by lalo-uy** [comment]
It will do the same , turn on the triac a % of time.
